### DISCUSSION TITLE
Should the way I am using Supabase Realtime here work?

### DISCUSSION THREAD

**Discussion by TomasHubelbauer** [supabase/supabase#17310] [Questions] [answered]
Hi,
I have a Supabase project with one table, `items`.
The table has RLS on with this policy:
```sql
CREATE POLICY "Enable access for authenticated users only" ON "public"."items"
AS PERMISSIVE FOR ALL
TO authenticated
USING (true)
WITH CHECK (true)
```

This means I can't use the public anon key to access anything in this table. (I hope so at least.)

I am using Supabase Auth to sign my user in and get a session so that I can access the data with my user.
I am using the `supabase-js` library which makes this very easy: after I call `supabase.auth.signInWithPassword` I can just start making calls like `supabase.from("items").select()` and they just work - the token is handled by the library and I don't need to do anything extra.

I enabled Realtime for the `items` table in the project settings under Replication hoping I can get the same ease of use here.

I am trying to use Supabase Realtime via the same `supabase-js` library. I don't want to use `realtime-js` because it shouldn't be necessary - the `supabase-js` Realtime API covers my needs. I am doing this:

```typescript
useEffect(() => {
  const channel = supabase
    .channel("items")
    .on(
      "postgres_changes",
      { event: "*", schema: "public", table: "items" },
      async (payload) => {
        console.log(payload);
      }
    )
    .subscribe();
  return () => {
    channel.unsubscribe();
  };
}, []);
```

Unfortunately, the `payload` callback is never called. I tried disabling RLS on my table to no avail.
The connection is made but I get no realtime data.

I checked to see if Supabase Realtime uses the token from Supabase Auth automatically the same way using Postgrest via `supabase-js` seems to do it. And I think that _is_ happening:

`supabase.realtime.accessToken` when logged is the same value as `supabase.auth.getSession().session?.access_token`.
I did however notice there are no tables in my `realtime` schema in the Supabase project.
I don't understand why the connection is completing (according to the `subscribe` callback) even when this is the case.

I also tried using `realtime-js` and passing in the token from the Supabase Auth session explicitly. This didn't work for me as I got a websocket connection refused error.

I have a feeling the token from Supabase Auth as used by Supabase Realtime via `supabase-js` is not in the format Supabase Realtime expects maybe? I saw discussions hinting at that but I honestly can't tell if that's my case or not.

I would expect this to work out of the box since the token seems to be passed from Supabase Auth to Supabase Realtime within `supabase-js`. Is this a fair expectation? Is this something that should work as of today?

Am I missing something?

**Comment by GaryAustin1** [comment]
You could try with no role in the policy to rule out your session on the client.
You can also do .subscribe(status=>console.log(status)) to get more info on the connection.
And there is https://realtime.supabase.co/inspector/new you could use to test if it is an instance setup issue or your code.

**Reply by TomasHubelbauer** [reply]
> You could try with no role in the policy to rule out your session on the client.

That means remove `TO authenticated`? Is this something I can do using the graphical policy editor? So if I understand correctly this is kind of turning off RLS without really turning off RLS, right? If this works that indicates the token Supabase Auth creates when I call `loginWithEmail` that's later used by Supabase Realtime doesn't have the right field for holding the value of the role?
```sql
CREATE POLICY "Enable access for authenticated users only" ON "public"."items"
AS PERMISSIVE FOR ALL
USING (true)
WITH CHECK (true)
```

> .You can also do `.subscribe(status=>console.log(status))` to get more info on the connection.

This always returns `SUBSCRIBED` even though I see no tables in the `realtime` schema after this. When I tried using `realtime-js` instead of `supabase-js` I would get a network error about the websocket connection being refused by the host. But with `supabase-js` and this code I posted this "works" in the sense that I don't get any errors or indications of any problems with the connection on the client, just no data.

> And there is https://realtime.supabase.co/inspector/new you could use to test if it is an instance setup issue or your code.

This seems extremely useful, thanks! I tried this and filled in:
- Checked "Enable DB changes" as that's what I'm interested in
- Filled in my project ref
- Had the host reflect the project ref correctly
- Filled in "test" for channel name - this can be just anything right?
- Left schema at "public"
- Filled in my table name "items"
- Left the filter empty
- Switched log level to "Debug"
- Added my public anon key to "Token"
- Left "Bearer" empty

The first issue I see is that the DB connection dashboard item looks like this:
<img width="372" alt="image" src="https://github.com/supabase/supabase/assets/6831144/1da787cc-fccf-4e57-abdf-2261653a5015">

Looks like it isn't subscribing to them despite my checking the related checkbox?
Nevertheless I tried changing a record in the DB and got no payloads to show up here.

I also tried using the `access_token` from my local storage's `sb-project-ref-auth-token` object to use the auth user's access token instead of the public anon access token. This made the Inspector modal freeze at the Connecting… step. This seems similar to what happened when I tried using `realtime-js` directly and passing in the access token from Supabase Auth only to get "websocket connection refused by the host" in the Network tab after doing that.

**Reply by TomasHubelbauer** [reply]
I should also mention I have Realtime enabled on the table in question and I do have the table enabled in the database replication settings.

**Reply by GaryAustin1** [reply]
supabase-js realtime works fine, I would not go with realtime-js.  You can test the anon role in RLS by just setting the role in the policy UI to nothing or anon.   Just trying to make sure your jwt/access token is actually set.

If your policy was TO authenticated then anon role would not work in your Bearer field.   Token is your anon or service role.  Bearer is your user jwt (not expired).

**Reply by TomasHubelbauer** [reply]
I set the RLS policy the way you suggest so that the table is public but the Inspector tool still shows Database - Not Subscribed even though I selected DB changes in the new connection modal. This is after setting the table to public and making sure I am using the public anon token in the new connection modal. This seems like a problem with Supabase Realtime or the Inspector since it is not coming from my code, right? But I am surprised I would be the only person affected by this so it probably still is somehow related to my setup.

Thanks for clearing up the difference between token and bearer.

I tried using my user's non-expired `access_token` from the local storage and placing it into the Bearer field (and clearing the Token field) and when I pressed Connect the button got stuck at the Connecting… state. I did not see anything in the Network tab in the Dev Tools though (XHR or other).

Am I using the Inspector incorrectly? I think definitely the DB changes not being listened to despite me checking the checkbox is suspect at least. The freeze when using the bearer, too.

Another question: in my own code, when using supabase-js with Realtime, am I responsible for getting my user's JWT and setting it as the bearer in the code? Or should that be something that happens automatically after I call `supabase.auth.signInWithEmail`?

**Reply by GaryAustin1** [reply]
Check the DB changes flag by clicking new connection.
Try with RLS off on the table.

I just verified the inspector works fine.

**Reply by TomasHubelbauer** [reply]
Okay, trying everything anew:
- I created a new table `tests` in the `public` schema with no RLS with `id` PK and a numeric `counter` column
- I opened the Inspector and filled this in exactly:
  - [x] Enable broadcast
  - [ ] Enable presence
  - [x] Enable DB changes
  - Project: `…` (my project ref)
  - Host: `https://xxuowuipbqlhewafcxgo.supabase.co`
  - Channel: `test` (I think this is an arbitrary name that doesn't need to be set up anywhere else?)
  - Schema: `public`
  - Table: `tests`
  - Filter: `` (no filter)
  - Log level: `debug`
  - Token: `…` (my public anon key)
  - Bearer: `` (empty - no `authorized` role user)

At this point the DB changes dashboard cell looks like this:
<img width="372" alt="image" src="https://github.com/supabase/supabase/assets/6831144/16cff8bb-3d32-400b-9cf2-b8297de23b3b">

I think this is definitely wrong at this point, right? RLS is off, the public anon key is correct, why wouldn't it subscribe?

Despite that I am making changes to the table in Supabase Dashboard. Tried an INSERT to give the table some data and an UPDATE by incrementing the `counter` column. Neither of these events showed up in the payload list:

<img width="869" alt="image" src="https://github.com/supabase/supabase/assets/6831144/5e679d37-8bca-4bec-ab06-875c5374e18f">

So I believe I am not connected with Supabase Realtime but this is the simplest setup I can think of, I am struggling to think of what it is that I could have messed up in this. :D

**Reply by GaryAustin1** [reply]
I personally would not put my anon and url (if they are real) on a public forum, especially in development with no RLS.

Did you enable realtime on tests?

**Reply by TomasHubelbauer** [reply]
I was under the impression both of these values are public (both of them is something that would end up in a browser bundle). But I will rotate them if they are not supposed to be shared. Realtime replication is enabled on `tests`.

**Reply by GaryAustin1** [reply]
They can be shared, but if you are in dev and don't have things locked down in RLS then anyone could access your instance.  I just would not put them in a major forum that gets scraped by Google and many people goto myself.

Don't know what else to tell you on your issue.  If you have RLS off and have all the Realtime settings (Update/Delete/Insert) selected and the table added that should be all you need.  I would expect an error if you are not getting subscribed.   RLS being wrong also should still subscribe and/or return an error.

Do you use Prisma?
Did you look in the realtime logs in the dashboard?
You might restart your instance or spin up another one to check.

**Reply by TomasHubelbauer** [reply]
Point taken about the key sharing, thank you. 😌 Alright, I think it is time I open a support ticket, not sure where to move from here myself. I don't use Prisma, just Next with the Supabase JS SDK. I wasn't aware of Realtime logs in the Dashboard, thanks for the tip. I do not see anything suspicious there at first glance, but I will ask SB support for what to look for there. And I'll do this after trying it all on a different project to make sure it is not a project-specific issue. I am using the hosted instance so I can't restart it, but I can rotate the project so I will do that. Thanks for your effort on this issue, I really appreciate the help!

**Reply by GaryAustin1** [reply]
![image](https://github.com/supabase/supabase/assets/54564956/b8dc2727-6e54-4f1a-a371-21e036b2ea7f)
If you are not in production.

**Reply by TomasHubelbauer** [reply, marked as answer]
FWIW this just didn't work for me for the longest time. I checked the Realtime Debugger in Studio when that was dropped on LWX and in there, everything worked fine. I've bumped the Supabase CLI and the Supabase JS SDK a few times since filing this and I just tried again and everything works as expected.
